compiler/src/iree/compiler/Codegen/Common/GPU/GPULowerToGlobalLoads.cpp

```diff
@@ -0,0 +1,338 @@
+  // workgroup memory. Assume there are N threads in the workgroup, then there
+  // are `num_subgroups = N / gpu.subgroup_size` subgroups in the workgroup.
+  //
+  // So we are slicing up the target memref into `num_subgroups` consecutive
+  // slices, and threads in the same subgroup will copy their slice to workgroup
+  // memory slice.
+
+  // Get the copy size:
+  auto copyMemRefType = cast<MemRefType>(copy.getOperand(1).getType());
+  if (!memref::isStaticShapeAndContiguousRowMajor(copyMemRefType)) {
+    copy.emitOpError("Cannot proceed: copy to non-static or non-contiguous, "
+                     "non-row major memref.");
+    return false;
+  }
+  auto rank = copyMemRefType.getRank();
+  SmallVector<OpFoldResult> tileSize(rank - 1, rewriter.getIndexAttr(1));
+
+  size_t elementBitWidth = copyMemRefType.getElementTypeBitWidth();
+  if (kNumBitsPerCopy % elementBitWidth != 0) {
+    copy.emitOpError("Cannot proceed: preferred copy size is not a multiple of "
+                     "element bit width.");
+    return false;
+  }
+  size_t elementsPerCopy = kNumBitsPerCopy / elementBitWidth;
+
+  // divide the copy by subgroup:
+  assert(subgroupSize.size() == 1); // only 1-D
+  assert(workgroupSize[0] % subgroupSize[0] == 0);
+
+  auto numSubgroups = workgroupSize[0] / subgroupSize[0];
+  auto totalCopySize = getTotalSize(copyMemRefType.getShape());
+  auto totalCopySizePerSubgroup = totalCopySize / numSubgroups;
+  auto numCopiesPerThread =
+      (totalCopySizePerSubgroup / elementsPerCopy) / subgroupSize[0];
+  auto residualElements =
+      totalCopySizePerSubgroup % (subgroupSize[0] * elementsPerCopy);
+
+  LDBG("-- elementsPerCopy: " << elementsPerCopy << "\n");
+  LDBG("-- workgroupSize: " << workgroupSize[0] << "\n");
+  LDBG("-- numSubgroups: " << numSubgroups << "\n");
+  LDBG("-- totalCopySize: " << totalCopySize << "\n");
+  LDBG("-- totalCopySizePerSubgroup: " << totalCopySizePerSubgroup << "\n");
+  LDBG("-- numCopiesPerThread: " << numCopiesPerThread << "\n");
+  LDBG("-- residualElements: " << residualElements << "\n");
+
+  if (residualElements != 0) {
+    copy.emitOpError(
+        "Cannot proceed: Cannot handle copying residual elements.");
+    return false;
+  }
+
+  // build For loop
+  SmallVector<Attribute> mapping;
+  for (int64_t i = 0, e = rank; i < e; ++i) {
+    unsigned mappingId = static_cast<unsigned>(gpu::MappingId::LinearDim0) + i;
+    mapping.push_back(gpu::GPUThreadMappingAttr::get(
+        context, static_cast<gpu::MappingId>(mappingId)));
+  }
+  std::reverse(mapping.begin(), mapping.end());
+
+  Value subgroupId = rewriter.create<gpu::SubgroupIdOp>(loc, nullptr);
+  Value laneId = rewriter.create<gpu::LaneIdOp>(loc, nullptr);
+
+  // TODO: make it multidimensional.
+
+  // compute number of loads per thread
+  auto sourceType = cast<MemRefType>(copy.getOperand(0).getType());
+  MemRefType localType = cast<MemRefType>(copy.getOutputs().front().getType());
```

Comment:
nit1: complete the sentence.
nit2: use `auto` because cast already spells the type.

```suggestion
  // Compute number of loads per thread.
  auto sourceType = cast<MemRefType>(copy.getOperand(0).getType());
  auot localType = cast<MemRefType>(copy.getOutputs().front().getType());
```

Reply:
bump, you probably missed this.